include/sparrow/variable_size_binary_layout.hpp

```diff
@@ -0,0 +1,307 @@
+     * layout.
+     *
+     * @tparam L the layout type
+     * @tparam is_const a boolean flag specifying whether this iterator is const.
+     */
+    template <class L, bool is_const>
+    class vs_binary_value_iterator : public iterator_base
+    <
+        vs_binary_value_iterator<L, is_const>,
+        mpl::constify_t<typename L::inner_value_type, is_const>,
+        std::contiguous_iterator_tag,
+        impl::get_inner_reference_t<L, is_const>
+    >
+    {
+    public:
+
+        using self_type = vs_binary_value_iterator<L, is_const>;
+        using base_type = iterator_base
+        <
+            self_type,
+            mpl::constify_t<typename L::inner_value_type, is_const>,
+            std::contiguous_iterator_tag,
+            impl::get_inner_reference_t<L, is_const>
+        >;
+        using reference = typename base_type::reference;
+        using difference_type = typename base_type::difference_type;
+        
+        using offset_iterator = std::conditional_t<
+            is_const, typename L::const_offset_iterator, typename L::offset_iterator
+        >;
+        using data_iterator = std::conditional_t<
+            is_const, typename L::const_data_iterator, typename L::data_iterator
+        >;
+
+        vs_binary_value_iterator() noexcept = default;
+        vs_binary_value_iterator(
+            offset_iterator offset_it,
+            data_iterator data_begin
+        );
+
+    private:
+
+        reference dereference() const;
+        void increment();
+        void decrement();
+        void advance(difference_type n);
+        difference_type distance_to(const self_type& rhs) const;
+        bool equal(const self_type& rhs) const;
+        bool less_than(const self_type& rhs) const;
+
+        offset_iterator m_offset_it;
+        const data_iterator m_data_begin;
+
+        friend class iterator_access;
+    };
+
+    /*
+     * @class variabe_size_binary_layout
+     *
+     * @brief Layout for arrays containing values consisting of a variable number of bytes.
+     *
+     * This layout is used to retrieve data in an array of values of a variable number of bytes
+     * (typically string objects). Values are stored contiguously in a data buffer, a single
+     * value is retrieved via an additional offset buffer, where each element is the beginning
+     * of the corresponinding value in the data buffer.
+     *
+     * Example:
+     *
+     * Let's consider the array ['please', 'allow', 'me', 'to', 'introduce', 'myself'].
+     * The internal buffers will be:
+     * - offset: [0, 6, 11, 13, 15, 24, 30]
+     * - data: ['p','l','e','a','s','e','a','l','l','o','w','m','e','t','o','i','n','t','r','o','d','u','c','e','m','y','s','e','l','f']
+     * 
+     * @tparam T the type of the data
+     * @tparam R the reference type to the data. This type is different form the reference type of the layout,
+     * which behaves like std::optional<R>.
```

Comment:
Consider adding a concept to constraint the reference type, as we know the kind of interface it is supposed to provide?

Reply:
We don't really know, except that an object T can be assign to an object R, the API really depends on T and can be anything.

Reply:
So it doesnt need to provide a way to know if there is a value, right?